org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/proxy/EclipseProxySettingsSupporter.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (c) 2015 the original author or authors.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     Ian Stewart-Binks (Red Hat, Inc.) - Bug 471943 - Make Buildship work behind the firewall
+ */
+
+package org.eclipse.buildship.core.proxy;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import com.google.common.base.Optional;
+import org.eclipse.core.net.proxy.IProxyData;
+import org.eclipse.buildship.core.CorePlugin;
+
+/**
+ * This class propagates Eclipse's proxy settings to System settings, and ensures that the existing System proxy settings
+ * are persisted.
+ */
+public class EclipseProxySettingsSupporter {
+
+    private static String savedHTTPProxyHost, savedHTTPProxyPort, savedHTTPProxyUser, savedHTTPProxyPassword;
+    private static final Lock lock = new ReentrantLock();
+
+    /**
+     * Configures the System properties proxy settings based on the Eclipse proxy settings.
+     */
+    public static void configureEclipseProxySettings() {
+        synchronized(EclipseProxySettingsSupporter.class) {
```

Comment:
I've attempted to make this block atomic. Suppose a thread, thread A, attempts to call `configureEclipseProxySettings`, and a thread, thread B, tries to do the same. Let thread A execute `storeSystemProxySettings` first, and acquire the lock in that method. Now run thread B. Thread B fails and does not get the lock. Thread B then continues on and executes `configureHTTPProxySettings`. Thread B updates the system settings with the Eclipse proxy settings. Thread A runs, and tries to store the default system settings, however Thread B just overwrote them.